### PR TITLE
Separating NEG controller from other controllers

### PR DESCRIPTION
**Pull request by sawsa307**

### CODE REVIEW

File: cmd/glbc/main.go [hunk 1]
```diff
@@ -235,46 +235,74 @@ func main() {
...
-			stopCh:      make(chan struct{}),
-			wg:          &sync.WaitGroup{},
-			leaderElect: false,
+	var wg sync.WaitGroup
+	if flags.F.EnableNEGController {
+		wg.Add(1)
+		defer wg.Done()
+		runNEG := func() {
+			runNEGController(ctx, rootLogger)
+			klog.Warning("NEG Controller exited.")
```
Comment (alexkats) [inline cmd/glbc/main.go:244]: I know that it's still in a draft state, but we have context logging at this point, so please change this and further calls to `klog` logger (apart from `Fatal`) to the context logger before merging. Thanks!

Comment (sawsa307) [reply cmd/glbc/main.go:244]: Yes I'll update this. Thanks for catching it!

File: cmd/glbc/main.go [hunk 2]
```diff
@@ -597,3 +570,23 @@ func collectLockAvailabilityMetrics(lockName, clusterType string, stopCh <-chan
 		}
 	}
 }
+
+func waitWithTimeout(wg *sync.WaitGroup, logger klog.Logger) {
+	doneCh := make(chan struct{})
+	go func() {
+		defer close(doneCh)
+		// Wait until all controllers are done with cleanup.
+		logger.Info("Starting to wait for NEG controller cleanup")
```
Comment (swetharepakula) [inline cmd/glbc/main.go:618]: The log lines need to be more generic -- we should make sure the passed in logger as the right context

**Review by swetharepakula — COMMENTED** [review]
For the first commit, can you provide more context as to why waitGroup and stopCh no longer need to be shared

File: cmd/glbc/main.go [hunk 3]
```diff
@@ -265,7 +265,7 @@ func main() {
 			runNEG = func() {
 				negElectionConfig, err := makeNEGLeaderElectionConfig(
 					ctx,
-					runOption{
+					leOption{
```
Comment (swetharepakula) [inline cmd/glbc/main.go:268]: nit, `leaderElectOption`  to be a little clearer.

File: cmd/glbc/main.go [hunk 4]
```diff
@@ -341,51 +364,36 @@ func makeLeaderElectionConfig(ctx *ingctx.ControllerContext, option runOption, l
...
-	wg := option.wg
-	var once sync.Once
-	// This ensures that stopCh is only closed once.
-	// Right now, we have two callers.
-	// One is triggered when the ASM configmap changes, and the other one is
-	// triggered by the SIGTERM handler.
-	closeStopCh := func() {
-		once.Do(func() { close(stopCh) })
-	}
+	logger = logger.WithValues("Running controllers", "All controllers other than NEG Controller")
```
Comment (swetharepakula) [inline cmd/glbc/main.go:367]: This is a little confusing. Perhaps we just add the function name into the logger and log a line with what is enabled

File: cmd/glbc/main.go [hunk 5]
```diff
@@ -253,41 +253,72 @@ func main() {
 	ctx := ingctx.NewControllerContext(kubeConfig, kubeClient, backendConfigClient, frontendConfigClient, firewallCRClient, svcNegClient, ingParamsClient, svcAttachmentClient, networkClient, eventRecorderKubeClient, cloud, namer, kubeSystemUID, ctxConfig, rootLogger)
 	go app.RunHTTPServer(ctx.HealthCheck, rootLogger)
 
-	if !flags.F.LeaderElection.LeaderElect {
```
[the remark this answers is not in the document]
Comment (sawsa307) [reply cmd/glbc/main.go:256]: Done.

**Review by swetharepakula — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**